### PR TITLE
Add support for new authentication type BASIC_AUTH_JENKINS_CSRF

### PR DESCRIPTION
**Pull request by a-breid-ericsson**
### Applicable Issues
Bad error handeling and not good enough user response when trying to access services with CSRF Protection.

### Description of the Change
 - EI now fetches Jenkins crumb only when BASIC_AUTH_JENKINS_CSRF
  is used.
- Better error handling caused by failure to connect to rest services

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by:

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -29,7 +29,7 @@
         <springLdapVersion>2.3.2.RELEASE</springLdapVersion>
         <springSecurityVersion>5.0.6.RELEASE</springSecurityVersion>
         <springSessionVersion>2.0.2.RELEASE</springSessionVersion>
-        <forkCountTests>2.0C</forkCountTests>
+        <forkCountTests>1</forkCountTests>
```
Comment (a-breid-ericsson) [inline pom.xml]: This will be changed to 2C before merge

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java [hunk 2]
```diff
@@ -188,100 +192,141 @@ private HttpHeaders prepareHeaders(String notificationMeta, JsonNode subscriptio
...
+            }
+            throw e;
         }
+    }
 
-        LOGGER.debug("Successfully fetched Jenkins crumb.");
-        JsonNode JenkinsJsonCrumbData = response.getBody();
-        return JenkinsJsonCrumbData;
+    /**
+     * This function replaes the user given context paths with the crumb issuer context path.
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: This method replaces

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java [hunk 3]
```diff
@@ -188,100 +192,141 @@ private HttpHeaders prepareHeaders(String notificationMeta, JsonNode subscriptio
...
+            throw new AuthorizationException(message, e);
+        } catch (HttpClientErrorException e) {
+            if (HttpStatus.UNAUTHORIZED == e.getStatusCode()) {
+                String message = "Failed to fetch crumb. Authentication failed, wrong username or password.";
+                LOGGER.error(message, e);
+                throw new AuthorizationException(message, e);
+            }
+            if (HttpStatus.NOT_FOUND == e.getStatusCode()) {
+                String message = String.format(
+                        "Failed to fetch crumb. The authentication type is %s, but SCRF Protection seems disabled in Jenkins.",
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: Isn't it CSRF?

File: src/main/java/com/ericsson/ei/exception/AuthorizationException.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+/*
+   Copyright 2017 Ericsson AB.
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/exception/AuthorizationException.java]: Update the year?

File: .travis.yml [hunk 5]
```diff
@@ -23,11 +23,9 @@ before_install:
 # This is only run before integrationTests job
 # To ensure docker containers are fully up and running we sleep 20s
```
Comment (e-pettersson-ericsson) [inline .travis.yml]: Since the sleep time has increased, update the comments to match :)

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java (hunk 2 above)
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: Done

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java (hunk 3 above)
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: Done

File: src/main/java/com/ericsson/ei/exception/AuthorizationException.java (hunk 4 above)
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/exception/AuthorizationException.java]: Done

File: .travis.yml (hunk 5 above)
Comment (a-breid-ericsson) [reply .travis.yml]: Done

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java [hunk 6]
```diff
@@ -123,168 +131,223 @@ public void informSubscriber(String aggregatedObject, JsonNode subscriptionJson)
...
+                LOGGER.error(message, e);
+                throw new AuthenticationException(message, e);
+            }
+            if (HttpStatus.NOT_FOUND == e.getStatusCode()) {
+                String message = String.format(
+                        "Failed to fetch crumb. The authentication type is %s,"
+                                + " but CSRF Protection seems disabled in Jenkins.",
+                        AUTHENTICATION_TYPE_BASIC_AUTH_JENKINS_CSRF);
+                LOGGER.warn(message, e);
+                return null;
```
Comment (evasiba-ericsson) [inline src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: this return will cancel the coming throw. Should be so?

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java [hunk 7]
```diff
@@ -123,168 +131,223 @@ public void informSubscriber(String aggregatedObject, JsonNode subscriptionJson)
...
-        }
-
         String username = getSubscriptionField("userName", subscriptionJson);
         String password = getSubscriptionField("password", subscriptionJson);
 
-        if (!username.equals("") && !password.equals("")) {
-            String encoding = Base64.getEncoder().encodeToString((username + ":" + password).getBytes());
-            headers.add("Authorization", "Basic " + encoding);
-            LOGGER.debug("Successfully added header for 'Authorization'");
+        boolean authorizationDetailsProvided = isAuthenticationDetailsProvided(authType, username,
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: I haven't followed the conversation about this very closely, but what was said about using authentication vs. authorization? :)

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java [hunk 8]
```diff
@@ -123,168 +131,223 @@ public void informSubscriber(String aggregatedObject, JsonNode subscriptionJson)
...
+        if (!authorizationDetailsProvided) {
+            return headers;
+        }
 
-            // Adding jenkins crumb if any
-            headers = addJenkinsCrumbData(headers, encoding, notificationMeta);
+        String encoding = Base64.getEncoder().encodeToString((username + ":" + password)
+                .getBytes());
+        headers.add("Authorization", "Basic " + encoding);
+        LOGGER.debug("Successfully added header for 'Authorization'");
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: Same comment regarding how we log authentication/authorization...  Isn't it more confusing if both are used interchangeably? Maybe stick to one both in documentation/logging and code unless we actually mean something different and want to emphasize this

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java (hunk 7 above)
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: Done

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java [hunk 9]
```diff
@@ -123,168 +131,223 @@ public void informSubscriber(String aggregatedObject, JsonNode subscriptionJson)
...
+        JsonNode jenkinsJsonCrumbData = fetchJenkinsCrumb(encoding, notificationMeta);
         if (jenkinsJsonCrumbData != null) {
             String crumbKey = jenkinsJsonCrumbData.get("crumbRequestField").asText();
             String crumbValue = jenkinsJsonCrumbData.get("crumb").asText();
             headers.add(crumbKey, crumbValue);
-            LOGGER.debug("Successfully added header: " + String.format("'%s':'%s'", crumbKey, crumbValue));
+            LOGGER.info("Successfully added header: " + String.format("'%s':'%s'", crumbKey,
+                    crumbValue));
+        } else {
+            LOGGER.info("Failed to fetch Jenkins Crumb data!");
```
Comment (e-pettersson-ericsson) [inline src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: Maybe clarify this log message a bit? Correct me if I'm wrong but this could happen when there is no CSRF enabled in Jenkins right? Then maybe rephrase the mesage so it is understood that there was no crumb to fetch but it's OK anyway.

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java (hunk 8 above)
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: No change due to header being named authorization

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java (hunk 6 above)
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: Yes, to ensure that subscriptions will still trigger if CSRF security is off.

File: src/main/java/com/ericsson/ei/subscription/InformSubscriber.java (hunk 9 above)
Comment (e-pettersson-ericsson) [reply src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: Or remove this else block entirely? Because the nested method fetchJenkinsCrumbIfAny logs all you need to know.

Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/subscription/InformSubscriber.java]: May remove this logging as it is logged as a warning in enother function.
